viogpu/viogpudo/viogpudo.cpp

```diff
@@ -3125,29 +3125,21 @@ VOID VioGpuAdapter::DpcRoutine(_In_ PDXGKRNL_INTERFACE pDxgkInterface)
                 DbgPrint(TRACE_LEVEL_VERBOSE, ("---> %s m_CtrlQueue pvbuf = %p len = %d\n", __FUNCTION__, pvbuf, len));
                 PGPU_CTRL_HDR pcmd = (PGPU_CTRL_HDR)pvbuf->buf;
                 PGPU_CTRL_HDR resp = (PGPU_CTRL_HDR)pvbuf->resp_buf;
-                PKEVENT evnt = pvbuf->event;
-                if (evnt == NULL)
-                {
-                    if (resp->type != VIRTIO_GPU_RESP_OK_NODATA)
-                    {
-                        DbgPrint(TRACE_LEVEL_ERROR, ("<--- %s type = %xlu flags = %lu fence_id = %llu ctx_id = %lu cmd_type = %lu\n",
-                            __FUNCTION__, resp->type, resp->flags, resp->fence_id, resp->ctx_id, pcmd->type));
-                    }
-                    m_CtrlQueue.ReleaseBuffer(pvbuf);
-                    continue;
+
+                if (resp->type >= VIRTIO_GPU_RESP_ERR_UNSPEC) {
+                   DbgPrint(TRACE_LEVEL_FATAL, ("!!!!! Command failed %d", resp->type));
                 }
-                switch (pcmd->type)
+                if (resp->type != VIRTIO_GPU_RESP_OK_NODATA)
                 {
-                case VIRTIO_GPU_CMD_GET_DISPLAY_INFO:
-                case VIRTIO_GPU_CMD_GET_EDID:
+                    DbgPrint(TRACE_LEVEL_ERROR, ("<--- %s type = %xlu flags = %lu fence_id = %llu ctx_id = %lu cmd_type = %lu\n",
+                        __FUNCTION__, resp->type, resp->flags, resp->fence_id, resp->ctx_id, pcmd->type));
+                }
+                if (pvbuf->complete_cb != NULL)
                 {
-                    ASSERT(evnt);
-                    KeSetEvent(evnt, IO_NO_INCREMENT, FALSE);
+                    pvbuf->complete_cb(pvbuf->complete_ctx);
                 }
-                break;
-                default:
-                    DbgPrint(TRACE_LEVEL_ERROR, ("<--- %s Unknown cmd type 0x%x\n", __FUNCTION__, resp->type));
-                    break;
+                if (!pvbuf->auto_release) {
```

Comment:
Looks like you forgot to remove `!` there.

Reply:
Oh no... It should be fixed now